Underscores ( _ ) can be in numbers. They don't change the meaning of the numbers, but improve readability

In [ ]:
# underscores in numbers
12000300
12_000_300

In [2]:
# scientific notation
# 3e4 = 3E4 = 3 * 10000 = 30000
# 3e-4 = 3E-4 = 3 * 1/10000 = 0.0003
print(0.000000000000000000000000003)

3e-27


In [ ]:
# octal numbers
# start with 0O or 0o (Digits must be from 0-7 only, 8 and 9 are not allowed)
print(0o123)

83


0o123 = (1×8²) + (2×8¹) + (3×8⁰)
      = (1×64) + (2×8) + (3×1)
      = 64 + 16 + 3
      = 83

In [4]:
# hexadecimal numbers
# start with 0X or 0x
print(0x123)

291


0x123 = (1×16²) + (2×16¹) + (3×16⁰)
      = (1×256) + (2×16) + (3×1)
      = 256 + 32 + 3
      = 291

Octal and hexadecimal are not widely used. Just remember octal starts with 0O or 0o, and hexadecimal start with 0X or 0x